File: platforms/windows/LuminousCosmic/Models/AstrologyModels.cs
namespace LuminousCosmic.Models;

// ═══════════════════════════════════════════════════════════════
// ENUMERATIONS
// ═══════════════════════════════════════════════════════════════

public enum ZodiacSign
{
    Aries, Taurus, Gemini, Cancer, Leo, Virgo,
    Libra, Scorpio, Sagittarius, Capricorn, Aquarius, Pisces
}

public enum Planet
{
    Sun, Moon, Mercury, Venus, Mars,
    Jupiter, Saturn, Uranus, Neptune, Pluto,
    NorthNode, SouthNode, Chiron, Ascendant, Midheaven
}

public enum HouseSystem
{
    Placidus, Koch, WholeSign, EqualHouse, Campanus, Regiomontanus
}

public enum AspectType
{
    Conjunction,    // 0 degrees
    Sextile,        // 60 degrees
    Square,         // 90 degrees
    Trine,          // 120 degrees
    Opposition,     // 180 degrees
    Quincunx,       // 150 degrees
    SemiSextile,    // 30 degrees
    SemiSquare,     // 45 degrees
    Sesquiquadrate, // 135 degrees
    Quintile        // 72 degrees
}

public enum Element
{
    Fire, Earth, Air, Water
}

public enum Modality
{
    Cardinal, Fixed, Mutable
}

public enum MoonPhase
{
    NewMoon, WaxingCrescent, FirstQuarter, WaxingGibbous,
    FullMoon, WaningGibbous, LastQuarter, WaningCrescent
}

public enum TransitType
{
    Conjunction, Opposition, Square, Trine, Sextile
}

// ═══════════════════════════════════════════════════════════════
// DATA MODELS
// ═══════════════════════════════════════════════════════════════

/// <summary>
/// Represents a planetary placement in the natal chart.
/// </summary>
public sealed class PlanetPlacement
{
    public Planet Planet { get; set; }
    public ZodiacSign Sign { get; set; }
    public int House { get; set; }
    public double Degrees { get; set; }        // 0-359.99 absolute ecliptic longitude
    public double SignDegrees { get; set; }     // 0-29.99 within sign
    public bool IsRetrograde { get; set; }

    public string DegreeDisplay =>
        $"{(int)SignDegrees}\u00b0{(int)((SignDegrees % 1) * 60)}' {Sign}";

    public string Glyph => GetPlanetGlyph(Planet);

    public static string GetPlanetGlyph(Planet planet) => planet switch
    {
        Planet.Sun => "\u2609",
        Planet.Moon => "\u263D",
        Planet.Mercury => "\u263F",
        Planet.Venus => "\u2640",
        Planet.Mars => "\u2642",
        Planet.Jupiter => "\u2643",
        Planet.Saturn => "\u2644",
        Planet.Uranus => "\u2645",
        Planet.Neptune => "\u2646",
        Planet.Pluto => "\u2647",
        Planet.NorthNode => "\u260A",
        Planet.SouthNode => "\u260B",
        Planet.Chiron => "\u26B7",
        Planet.Ascendant => "AC",
        Planet.Midheaven => "MC",
        _ => "?"
    };
}

/// <summary>
/// Represents an aspect (angular relationship) between two planets.
/// </summary>
public sealed class Aspect
{
    public Planet Planet1 { get; set; }
    public Planet Planet2 { get; set; }
    public AspectType Type { get; set; }
    public double Orb { get; set; }            // Deviation from exact aspect
    public bool IsApplying { get; set; }

    public string Description => $"{Planet1} {TypeSymbol} {Planet2} ({Orb:F1}\u00b0)";

    public string TypeSymbol => Type switch
    {
        AspectType.Conjunction => "\u260C",
        AspectType.Sextile => "\u26B9",
        AspectType.Square => "\u25A1",
        AspectType.Trine => "\u25B3",
        AspectType.Opposition => "\u260D",
        AspectType.Quincunx => "Qx",
        _ => "\u2022"
    };
}

/// <summary>
/// Represents a house cusp in the natal chart.
/// </summary>
public sealed class HouseCusp
{
    public int HouseNumber { get; set; }       // 1-12
    public ZodiacSign Sign { get; set; }
    public double Degrees { get; set; }        // Absolute ecliptic longitude
    public double SignDegrees { get; set; }

    public string Label => HouseNumber switch
    {
        1 => "I - Self",
        2 => "II - Resources",
        3 => "III - Communication",
        4 => "IV - Home",
        5 => "V - Creativity",
        6 => "VI - Service",
        7 => "VII - Partnership",
        8 => "VIII - Transformation",
        9 => "IX - Philosophy",
        10 => "X - Career",
        11 => "XI - Community",
        12 => "XII - Transcendence",
        _ => $"House {HouseNumber}"
    };
}

/// <summary>
/// Birth data required to calculate a natal chart.
/// </summary>
public sealed class BirthData
{
    public string Name { get; set; } = string.Empty;
    public DateTime BirthDate { get; set; } = DateTime.Now;
    public TimeSpan BirthTime { get; set; } = TimeSpan.Zero;
    public string BirthCity { get; set; } = string.Empty;
    public double Latitude { get; set; }
    public double Longitude { get; set; }
    public double TimezoneOffset { get; set; }

    public DateTime BirthDateTime => BirthDate.Date + BirthTime;
}

/// <summary>
/// A complete natal chart with all calculated positions.
/// </summary>
public sealed class NatalChart
{
    public BirthData BirthData { get; set; } = new();
    public List<PlanetPlacement> Planets { get; set; } = new();
    public List<HouseCusp> Houses { get; set; } = new();
    public List<Aspect> Aspects { get; set; } = new();
    public HouseSystem HouseSystem { get; set; } = HouseSystem.Placidus;

    public PlanetPlacement? SunSign => Planets.FirstOrDefault(p => p.Planet == Planet.Sun);
    public PlanetPlacement? MoonSign => Planets.FirstOrDefault(p => p.Planet == Planet.Moon);
    public PlanetPlacement? Rising => Planets.FirstOrDefault(p => p.Planet == Planet.Ascendant);

    public Element DominantElement
    {
        get
        {
            var counts = Planets
                .Where(p => p.Planet != Planet.Ascendant && p.Planet != Planet.Midheaven)
                .GroupBy(p => GetElement(p.Sign))
                .OrderByDescending(g => g.Count())
                .FirstOrDefault();
            return counts?.Key ?? Element.Earth;
        }
    }

    public static Element GetElement(ZodiacSign sign) => sign switch
    {
        ZodiacSign.Aries or ZodiacSign.Leo or ZodiacSign.Sagittarius => Element.Fire,
        ZodiacSign.Taurus or ZodiacSign.Virgo or ZodiacSign.Capricorn => Element.Earth,
        ZodiacSign.Gemini or ZodiacSign.Libra or ZodiacSign.Aquarius => Element.Air,
        ZodiacSign.Cancer or ZodiacSign.Scorpio or ZodiacSign.Pisces => Element.Water,
        _ => Element.Earth
    };

    public static Modality GetModality(ZodiacSign sign) => sign switch
    {
        ZodiacSign.Aries or ZodiacSign.Cancer or ZodiacSign.Libra or ZodiacSign.Capricorn => Modality.Cardinal,
        ZodiacSign.Taurus or ZodiacSign.Leo or ZodiacSign.Scorpio or ZodiacSign.Aquarius => Modality.Fixed,
        ZodiacSign.Gemini or ZodiacSign.Virgo or ZodiacSign.Sagittarius or ZodiacSign.Pisces => Modality.Mutable,
        _ => Modality.Cardinal
    };
}

/// <summary>
/// Represents a current planetary transit affecting the natal chart.
/// </summary>
public sealed class Transit
{
    public Planet TransitingPlanet { get; set; }
    public Planet NatalPlanet { get; set; }
    public TransitType Type { get; set; }
    public ZodiacSign TransitSign { get; set; }
    public double TransitDegrees { get; set; }
    public DateTime ExactDate { get; set; }
    public bool IsActive { get; set; }
    public string Interpretation { get; set; } = string.Empty;
    public string Theme { get; set; } = string.Empty;

    public string Summary =>
        $"{PlanetPlacement.GetPlanetGlyph(TransitingPlanet)} {TransitingPlanet} " +
        $"{Type} natal {PlanetPlacement.GetPlanetGlyph(NatalPlanet)} {NatalPlanet}";
}

/// <summary>
/// Moon phase information for a given date.
/// </summary>
public sealed class MoonPhaseInfo
{
    public MoonPhase Phase { get; set; }
    public ZodiacSign MoonSign { get; set; }
    public double Illumination { get; set; }     // 0.0 - 1.0
    public DateTime NextPhaseDate { get; set; }

    public string PhaseName => Phase switch
    {
        MoonPhase.NewMoon => "New Moon",
        MoonPhase.WaxingCrescent => "Waxing Crescent",
        MoonPhase.FirstQuarter => "First Quarter",
        MoonPhase.WaxingGibbous => "Waxing Gibbous",
        MoonPhase.FullMoon => "Full Moon",
        MoonPhase.WaningGibbous => "Waning Gibbous",
        MoonPhase.LastQuarter => "Last Quarter",
        MoonPhase.WaningCrescent => "Waning Crescent",
        _ => "Unknown"
    };

    public string PhaseEmoji => Phase switch
    {
        MoonPhase.NewMoon => "\U0001F311",
        MoonPhase.WaxingCrescent => "\U0001F312",
        MoonPhase.FirstQuarter => "\U0001F313",
        MoonPhase.WaxingGibbous => "\U0001F314",
        MoonPhase.FullMoon => "\U0001F315",
        MoonPhase.WaningGibbous => "\U0001F316",
        MoonPhase.LastQuarter => "\U0001F317",
        MoonPhase.WaningCrescent => "\U0001F318",
        _ => "\U0001F311"
    };

    public string Guidance => Phase switch
    {
        MoonPhase.NewMoon => "Set intentions. Plant seeds of new beginnings.",
        MoonPhase.WaxingCrescent => "Take the first steps. Build momentum.",
        MoonPhase.FirstQuarter => "Take decisive action. Overcome obstacles.",
        MoonPhase.WaxingGibbous => "Refine your approach. Trust the process.",
        MoonPhase.FullMoon => "Celebrate culmination. Release what no longer serves.",
        MoonPhase.WaningGibbous => "Share wisdom. Express gratitude.",
        MoonPhase.LastQuarter => "Let go. Forgive. Make space for the new.",
        MoonPhase.WaningCrescent => "Rest. Reflect. Surrender to stillness.",
        _ => "Observe the cosmic rhythm."
    };
}

/// <summary>
/// Daily reflection / journal entry.
/// </summary>
public sealed class ReflectionEntry
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public DateTime Date { get; set; } = DateTime.Today;
    public string Prompt { get; set; } = string.Empty;
    public string Response { get; set; } = string.Empty;
    public MoonPhase MoonPhase { get; set; }
    public ZodiacSign MoonSign { get; set; }
    public List<string> Tags { get; set; } = new();
    public int MoodRating { get; set; }           // 1-5
}

/// <summary>
/// A chapter in the cosmic architecture library.
/// </summary>
public sealed class CosmicChapter
{
    public int Number { get; set; }
    public string Title { get; set; } = string.Empty;
    public string Subtitle { get; set; } = string.Empty;
    public string Content { get; set; } = string.Empty;
    public string Category { get; set; } = string.Empty;
    public ZodiacSign? AssociatedSign { get; set; }
    public Planet? AssociatedPlanet { get; set; }
    public bool IsUnlocked { get; set; } = true;
    public double ReadingProgress { get; set; }   // 0.0 - 1.0
}

/// <summary>
/// Guided meditation session data.
/// </summary>
public sealed class MeditationSession
{
    public string Title { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public TimeSpan Duration { get; set; }
    public string Theme { get; set; } = string.Empty;
    public ZodiacSign? AssociatedSign { get; set; }
    public Planet? AssociatedPlanet { get; set; }
    public List<string> Prompts { get; set; } = new();
}

/// <summary>
/// Static zodiac sign metadata.
/// </summary>
public static class ZodiacMetadata
{
    public static string GetGlyph(ZodiacSign sign) => sign switch
    {
        ZodiacSign.Aries => "\u2648",
        ZodiacSign.Taurus => "\u2649",
        ZodiacSign.Gemini => "\u264A",
        ZodiacSign.Cancer => "\u264B",
        ZodiacSign.Leo => "\u264C",
        ZodiacSign.Virgo => "\u264D",
        ZodiacSign.Libra => "\u264E",
        ZodiacSign.Scorpio => "\u264F",
        ZodiacSign.Sagittarius => "\u2650",
        ZodiacSign.Capricorn => "\u2651",
        ZodiacSign.Aquarius => "\u2652",
        ZodiacSign.Pisces => "\u2653",
        _ => "?"
    };

    public static string GetRuler(ZodiacSign sign) => sign switch
    {
        ZodiacSign.Aries => "Mars",
        ZodiacSign.Taurus => "Venus",
        ZodiacSign.Gemini => "Mercury",
        ZodiacSign.Cancer => "Moon",
        ZodiacSign.Leo => "Sun",
        ZodiacSign.Virgo => "Mercury",
        ZodiacSign.Libra => "Venus",
        ZodiacSign.Scorpio => "Pluto",
        ZodiacSign.Sagittarius => "Jupiter",
        ZodiacSign.Capricorn => "Saturn",
        ZodiacSign.Aquarius => "Uranus",
        ZodiacSign.Pisces => "Neptune",
        _ => "Unknown"
    };

    public static string GetDateRange(ZodiacSign sign) => sign switch
    {
        ZodiacSign.Aries => "Mar 21 - Apr 19",
        ZodiacSign.Taurus => "Apr 20 - May 20",
        ZodiacSign.Gemini => "May 21 - Jun 20",
        ZodiacSign.Cancer => "Jun 21 - Jul 22",
        ZodiacSign.Leo => "Jul 23 - Aug 22",
        ZodiacSign.Virgo => "Aug 23 - Sep 22",
        ZodiacSign.Libra => "Sep 23 - Oct 22",
        ZodiacSign.Scorpio => "Oct 23 - Nov 21",
        ZodiacSign.Sagittarius => "Nov 22 - Dec 21",
        ZodiacSign.Capricorn => "Dec 22 - Jan 19",
        ZodiacSign.Aquarius => "Jan 20 - Feb 18",
        ZodiacSign.Pisces => "Feb 19 - Mar 20",
        _ => ""
    };
}

File: platforms/android/app/src/main/java/com/luminous/cosmic/data/models/AstrologyModels.kt
package com.luminous.cosmic.data.models

import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime

// ─────────────────────────────────────────────
// Zodiac & Celestial Enumerations
// ─────────────────────────────────────────────

enum class ZodiacSign(
    val symbol: String,
    val unicode: String,
    val element: Element,
    val modality: Modality,
    val rulingPlanet: Planet,
    val degreesStart: Float
) {
    ARIES("Aries", "\u2648", Element.FIRE, Modality.CARDINAL, Planet.MARS, 0f),
    TAURUS("Taurus", "\u2649", Element.EARTH, Modality.FIXED, Planet.VENUS, 30f),
    GEMINI("Gemini", "\u264A", Element.AIR, Modality.MUTABLE, Planet.MERCURY, 60f),
    CANCER("Cancer", "\u264B", Element.WATER, Modality.CARDINAL, Planet.MOON, 90f),
    LEO("Leo", "\u264C", Element.FIRE, Modality.FIXED, Planet.SUN, 120f),
    VIRGO("Virgo", "\u264D", Element.EARTH, Modality.MUTABLE, Planet.MERCURY, 150f),
    LIBRA("Libra", "\u264E", Element.AIR, Modality.CARDINAL, Planet.VENUS, 180f),
    SCORPIO("Scorpio", "\u264F", Element.WATER, Modality.FIXED, Planet.PLUTO, 210f),
    SAGITTARIUS("Sagittarius", "\u2650", Element.FIRE, Modality.MUTABLE, Planet.JUPITER, 240f),
    CAPRICORN("Capricorn", "\u2651", Element.EARTH, Modality.CARDINAL, Planet.SATURN, 270f),
    AQUARIUS("Aquarius", "\u2652", Element.AIR, Modality.FIXED, Planet.URANUS, 300f),
    PISCES("Pisces", "\u2653", Element.WATER, Modality.MUTABLE, Planet.NEPTUNE, 330f);

    companion object {
        fun fromDegree(degree: Float): ZodiacSign {
            val normalized = ((degree % 360f) + 360f) % 360f
            return entries.lastOrNull { normalized >= it.degreesStart } ?: ARIES
        }
    }
}

enum class Element(val displayName: String) {
    FIRE("Fire"), EARTH("Earth"), AIR("Air"), WATER("Water")
}

enum class Modality(val displayName: String) {
    CARDINAL("Cardinal"), FIXED("Fixed"), MUTABLE("Mutable")
}

enum class Planet(
    val symbol: String,
    val unicode: String,
    val orbDefault: Float
) {
    SUN("Sun", "\u2609", 10f),
    MOON("Moon", "\u263D", 10f),
    MERCURY("Mercury", "\u263F", 7f),
    VENUS("Venus", "\u2640", 7f),
    MARS("Mars", "\u2642", 7f),
    JUPITER("Jupiter", "\u2643", 9f),
    SATURN("Saturn", "\u2644", 9f),
    URANUS("Uranus", "\u2645", 5f),
    NEPTUNE("Neptune", "\u2646", 5f),
    PLUTO("Pluto", "\u2647", 5f),
    NORTH_NODE("North Node", "\u260A", 5f),
    CHIRON("Chiron", "\u26B7", 3f);
}

enum class House(val number: Int, val description: String) {
    FIRST(1, "Self & Identity"),
    SECOND(2, "Values & Possessions"),
    THIRD(3, "Communication & Siblings"),
    FOURTH(4, "Home & Roots"),
    FIFTH(5, "Creativity & Romance"),
    SIXTH(6, "Health & Service"),
    SEVENTH(7, "Partnerships"),
    EIGHTH(8, "Transformation & Shared Resources"),
    NINTH(9, "Philosophy & Travel"),
    TENTH(10, "Career & Public Image"),
    ELEVENTH(11, "Community & Aspirations"),
    TWELFTH(12, "Unconscious & Spirituality")
}

enum class AspectType(
    val symbol: String,
    val angle: Float,
    val orb: Float,
    val nature: AspectNature
) {
    CONJUNCTION("\u260C", 0f, 8f, AspectNature.MAJOR),
    SEXTILE("\u26B9", 60f, 4f, AspectNature.HARMONIOUS),
    SQUARE("\u25A1", 90f, 7f, AspectNature.CHALLENGING),
    TRINE("\u25B3", 120f, 7f, AspectNature.HARMONIOUS),
    OPPOSITION("\u260D", 180f, 8f, AspectNature.CHALLENGING),
    QUINCUNX("Qx", 150f, 3f, AspectNature.MINOR),
    SEMISEXTILE("SSx", 30f, 2f, AspectNature.MINOR);
}

enum class AspectNature {
    MAJOR, HARMONIOUS, CHALLENGING, MINOR
}

enum class MoonPhase(val displayName: String, val emoji: String) {
    NEW_MOON("New Moon", "\uD83C\uDF11"),
    WAXING_CRESCENT("Waxing Crescent", "\uD83C\uDF12"),
    FIRST_QUARTER("First Quarter", "\uD83C\uDF13"),
    WAXING_GIBBOUS("Waxing Gibbous", "\uD83C\uDF14"),
    FULL_MOON("Full Moon", "\uD83C\uDF15"),
    WANING_GIBBOUS("Waning Gibbous", "\uD83C\uDF16"),
    LAST_QUARTER("Last Quarter", "\uD83C\uDF17"),
    WANING_CRESCENT("Waning Crescent", "\uD83C\uDF18")
}

// ─────────────────────────────────────────────
// Data Models
// ─────────────────────────────────────────────

data class BirthData(
    val name: String = "",
    val birthDate: LocalDate = LocalDate.of(1990, 1, 1),
    val birthTime: LocalTime = LocalTime.NOON,
    val birthPlace: String = "",
    val latitude: Double = 0.0,
    val longitude: Double = 0.0,
    val timeZoneOffset: Int = 0
)

data class PlanetPlacement(
    val planet: Planet,
    val sign: ZodiacSign,
    val house: House,
    val degree: Float,
    val isRetrograde: Boolean = false
) {
    val formattedDegree: String
        get() {
            val signDegree = degree % 30f
            val deg = signDegree.toInt()
            val min = ((signDegree - deg) * 60).toInt()
            return "${sign.symbol} $deg\u00B0$min'"
        }
}

data class HouseCusp(
    val house: House,
    val sign: ZodiacSign,
    val degree: Float
)

data class Aspect(
    val planet1: Planet,
    val planet2: Planet,
    val type: AspectType,
    val orb: Float,
    val isApplying: Boolean = false
)

data class NatalChart(
    val birthData: BirthData,
    val planets: List<PlanetPlacement>,
    val houses: List<HouseCusp>,
    val aspects: List<Aspect>,
    val ascendantDegree: Float,
    val midheavenDegree: Float
) {
    val sunSign: ZodiacSign
        get() = planets.first { it.planet == Planet.SUN }.sign
    val moonSign: ZodiacSign
        get() = planets.first { it.planet == Planet.MOON }.sign
    val risingSign: ZodiacSign
        get() = ZodiacSign.fromDegree(ascendantDegree)
}

data class Transit(
    val planet: Planet,
    val natalPlanet: Planet,
    val aspectType: AspectType,
    val description: String,
    val startDate: LocalDate,
    val exactDate: LocalDate,
    val endDate: LocalDate,
    val intensity: Float = 0.5f
)

data class DailyInsight(
    val date: LocalDate,
    val title: String,
    val body: String,
    val moonPhase: MoonPhase,
    val transits: List<Transit>,
    val affirmation: String,
    val focusArea: String
)

data class JournalEntry(
    val id: String = java.util.UUID.randomUUID().toString(),
    val date: LocalDateTime = LocalDateTime.now(),
    val prompt: String = "",
    val content: String = "",
    val mood: String = "",
    val tags: List<String> = emptyList()
)

data class Meditation(
    val id: String,
    val title: String,
    val subtitle: String,
    val description: String,
    val durationMinutes: Int,
    val category: MeditationCategory,
    val steps: List<MeditationStep>
)

data class MeditationStep(
    val title: String,
    val instruction: String,
    val durationSeconds: Int,
    val breathPattern: BreathPattern? = null
)

data class BreathPattern(
    val inhaleSeconds: Int,
    val holdSeconds: Int,
    val exhaleSeconds: Int
)

enum class MeditationCategory(val displayName: String) {
    STARGAZER("Stargazer's Attunement"),
    LUNAR("Lunar Alignment"),
    ELEMENTAL("Elemental Balance"),
    PLANETARY("Planetary Meditation")
}

data class Chapter(
    val id: String,
    val number: Int,
    val title: String,
    val subtitle: String,
    val description: String,
    val sections: List<ChapterSection>,
    val isUnlocked: Boolean = true
)

data class ChapterSection(
    val title: String,
    val content: String,
    val relatedSign: ZodiacSign? = null
)

// ─────────────────────────────────────────────
// Navigation
// ─────────────────────────────────────────────

sealed class Screen(val route: String) {
    object Onboarding : Screen("onboarding")
    object Dashboard : Screen("dashboard")
    object NatalChart : Screen("natal_chart")
    object DailyReflection : Screen("daily_reflection")
    object Meditation : Screen("meditation")
    object ChapterLibrary : Screen("chapter_library")
}
